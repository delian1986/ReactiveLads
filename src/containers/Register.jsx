import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { RegisterForm as Register } from "../components/user/RegisterForm";
import { registerAsync } from "../store/actions/auth";
import { getIsAuthPending, getMessage } from "../store/selectors";

const mapStateToProps = (state) => {
  return {
    message: getMessage(state),
    isPending: getIsAuthPending(state)
  };
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators(
    {
      register: ({ email, password }) => registerAsync({ email, password })
    },
    dispatch
  );
};
export default connect(mapStateToProps, mapDispatchToProps)(Register);
