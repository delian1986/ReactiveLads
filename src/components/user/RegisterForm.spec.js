import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import Register from "../../containers/Register";
import { mount } from "enzyme";

const mockStore = configureStore([]);

describe("tests <RegisterFrom/> component", () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({
      auth: {
        isPending: false
      },
      message: {}
    });
    store.dispatch = jest.fn();
    component = renderer.create(
      <Provider store={store}>
        <Register />
      </Provider>
    );
  });
  describe("Test rendering component", () => {
    it("should render with given state from Redux store", () => {
      expect(component.toJSON()).toMatchSnapshot();
    });

    it("should render three input fields", function () {
      expect(component.root.findAllByType("input").length).toEqual(3);
    });
  });

  describe("Actions", () => {
    it("should dispatch an action on button click", () => {
      const fakeEvent = { preventDefault: () => true };
      renderer.act(() => {
        component.root.findByType("form").props.onSubmit(fakeEvent);
      });

      expect(store.dispatch).toHaveBeenCalledTimes(1);
    });
  });

  describe("integration test", () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <Register />
        </Provider>
      );
    });

    it("button should unlock on valid inputs", function () {
      const button = wrapper.find("button");
      expect(button.props().disabled).toEqual(true);

      const emailField = wrapper.find('input[name="email"]');
      emailField.simulate("change", { target: { value: "123@123.bg" } });
      expect(wrapper.find("button").props().disabled).toEqual(true);

      const passwordField = wrapper.find('input[name="password"]');
      passwordField.simulate("change", { target: { value: "1233" } });
      expect(wrapper.find("button").props().disabled).toEqual(true);

      const repeatPassword = wrapper.find('input[name="repeatPassword"]');
      repeatPassword.simulate("change", { target: { value: "123" } });
      expect(wrapper.find("button").props().disabled).toEqual(true);

      repeatPassword.simulate("change", { target: { value: "1233" } });
      expect(wrapper.find("button").props().disabled).toEqual(false);
    });

    it("button should stay locked on invalid inputs", function () {
      const button = wrapper.find("button");
      expect(button.props().disabled).toEqual(true);

      const emailField = wrapper.find('input[name="email"]');
      emailField.simulate("change", { target: { value: "fail@test" } });
      expect(wrapper.find("button").props().disabled).toEqual(true);

      const repeatPassword = wrapper.find('input[name="password"]');
      repeatPassword.simulate("change", { target: { value: "1233" } });

      const passwordField = wrapper.find('input[name="password"]');
      passwordField.simulate("change", { target: { value: "1233" } });
      expect(wrapper.find("button").props().disabled).toEqual(true);
    });
  });
});
