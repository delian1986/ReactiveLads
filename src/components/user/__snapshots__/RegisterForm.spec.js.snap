// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests <RegisterFrom/> component Test rendering component should render with given state from Redux store 1`] = `
<div
  className="container h-100"
>
  <div
    className="row h-100 pt-3 justify-content-center"
  >
    <div
      className="col-6 mx-auto"
    >
      <div
        className="card card-body bg-light"
      >
        <form
          onSubmit={[Function]}
        >
          <h2>
            Register
          </h2>
          <div
            className="form-group"
          >
            <label
              htmlFor="email"
            >
              Email
            </label>
            <input
              className="form-control"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Your email here..."
              type="email"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <label
              htmlFor="password"
            >
              Password
            </label>
            <input
              className="form-control"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Password 4 chars min..."
              type="password"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <label
              htmlFor="repeatPassword"
            >
              Repeat Password
            </label>
            <input
              className="form-control"
              name="repeatPassword"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Repeat your password here..."
              type="password"
              value=""
            />
          </div>
          <button
            className="btn btn-block"
            disabled={true}
            type="submit"
          >
            Register now !
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
